pyproject.toml

```diff
@@ -276,3 +275,27 @@ lint.ignore = [
 
 [tool.ruff.lint.pydocstyle]
 convention = "pep257"
+
+[tool.pytest.ini_options]
```

Comment:
> I don't think we can move these here yet.
> Downstreams (including ours) generally run their unit tests against installed packages, not using tox. Focal's pytest doesn't support putting these options here.

Correct - on Focal pytest will continue to behave as it did before, and on >=Jammy pytest will start using these values. What's the problem with that? I can see an argument for dropping some of these values - the `testpaths` value is obviously unnecessary, but I don't think that any of these values are harmful, are they?

Reply:
Why will Focal pytest continue to behave as it did before? These used to be defined in tox.ini, which Focal's python will respect, but now they live in pyproject.toml, which Focal's pytest will not respect. I'm saying that I think the options need to continue to live in tox.ini until downstreams have caught up to allowing this configuration in pyproject.toml.

Reply:
> Why will Focal pytest continue to behave as it did before?

Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now. Nothing changes on Focal. And even if it did I don't think it would matter. These aren't even necessary, they are just reasonable defaults to have when debugging.

Reply:
> Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now.

This isn't true. [Pytest uses tox.ini](https://docs.pytest.org/en/7.1.x/reference/customize.html#tox-ini) regardless of the existence of tox. It is a configuration file for pytest completely independent of tox.

>  These aren't even necessary, they are just reasonable defaults to have when debugging.

I don't think this is true. Without these settings we'll run integration tests by default when invoking pytest. Do we have something telling pybuild to not do that? We'll also get warnings about custom marks, but I suppose that isn't the end of the world.

Reply:
> > Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now.
> 
> This isn't true. [Pytest uses tox.ini](https://docs.pytest.org/en/7.1.x/reference/customize.html#tox-ini) regardless of the existence of tox. It is a configuration file for pytest completely independent of tox.

Huh, weird

> > These aren't even necessary, they are just reasonable defaults to have when debugging.
> 
> I don't think this is true.

The lowest-supported ci test [passes with this configuration](https://github.com/canonical/cloud-init/actions/runs/10378310058/job/28734306492?pr=5607) in pytest:

```sh
+ tox -e lowest-supported
lowest-supported: install_deps> python -I -m pip install configobj==5.0.6 jinja2==2.10.1 jsonpatch==1.23 jsonschema==3.2.0 markupsafe==2.0.1 oauthlib==3.1.0 passlib pyserial==3.4 pytest-cov==2.8.1 pytest-mock==1.10.4 pytest==4.6.9 pyyaml==5.3.1 requests==2.22.0 responses==0.9.0 setuptools==45.2.0
lowest-supported: freeze> python -m pip freeze --all
lowest-supported: atomicwrites==1.4.1,attrs==24.2.0,certifi==2024.7.4,chardet==3.0.4,configobj==5.0.6,cookies==2.2.1,coverage==7.6.1,idna==2.8,Jinja2==2.10.1,jsonpatch==1.23,jsonpointer==3.0.0,jsonschema==3.2.0,MarkupSafe==2.0.1,more-itertools==10.4.0,oauthlib==3.1.0,packaging==24.1,passlib==1.7.4,pip==24.1,pluggy==0.13.1,py==1.11.0,pyrsistent==0.20.0,pyserial==3.4,pytest==4.6.9,pytest-cov==2.8.1,pytest-mock==1.10.4,PyYAML==5.3.1,requests==2.22.0,responses==0.9.0,setuptools==45.2.0,six==1.16.0,urllib3==1.25.11,wcwidth==0.2.13,wheel==0.43.0
lowest-supported: commands[0]> .tox/lowest-supported/bin/python -m pytest -m 'not hypothesis_slow' --cov=cloud-init --cov-branch tests/unittests
```

Note that the options defined in in `pyproject.toml` somehow got invoked, which is a bit mysterious, but presumably that's some sort of tox magic.

> Without these settings we'll run integration tests by default when invoking pytest. Do we have something telling pybuild to not do that? We'll also get warnings about custom marks, but I suppose that isn't the end of the world.

Our packaging invokes pytest using `make check`, which runs `python3 -m pytest -v tests/unittests cloudinit`.

So yeah, we'll get some warnings about custom marks, but that should be it. We're also used to ignoring deprecation warnings and so I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that.

Reply:
> I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that

I would prefer it. We don't **need** any of those pytest options, but they do add convenience. If we ever find we do need to update these options such that they are required on focal, that would require either migrating all of the options back to tox.ini or managing two configuration files. Additionally, I know of at least one downstream where they write and test code on the OS they're wanting to patch, and so I think it would nice to allow them to keep things as "old version compatible" as possible. 

If this required us to trade some kind of functionality I would have a different opinion, but AFAIK there's really no difference between putting pytest configuration in tox.ini vs pyproject.toml.

Reply:
> > I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that
> 
> I would prefer it.

Sounds good, +1